Tweaks for "Giobonvi development" PR, and a lot of other loosely connected commits

rowanthorpe opened the pull request:
See my review on #72 (which this PR is against for the changes) for explanation about the commits.

rowanthorpe commented:
BTW: I tested these changes with `test()` and `unitTests()`, and read through the results of running jsdoc on `code.gs` and `unit-tests.gs` (which looked fine).

rowanthorpe commented:
I just added a commit to add a small sleep to the fetch-retry (instant-retry will more often than not fail just like the first one). The function is as-per [this documentation](https://developers.google.com/apps-script/reference/utilities/utilities#sleep(Integer)). I ran `test()` with that included, no syntax issues, etc. Didn't try editing it to cause deliberate fetch-failures to test the actual logic yet. Will try later if I find time.

rowanthorpe commented:
@GioBonvi regrading mistyped type: Oops, stupid oversight from me there - sorry, did it in a hurry. Fixed it now.

GioBonvi commented:
Thank you @rowanthorpe:

First of all I want to explain that I actually did not try using JSDoc to build the documentation; in fact I decided to adopt it because:

- I had thought about it, but never had the right incentive to implement it;
- @baatochan used it in #59;
- I discovered that my editor (as well as other ones) can use JSDoc comments to track variable types, to make the autocompletion better and give parameters descriptions as well as other features;

So, regarding HTML/Markdown I'd go with markdown since:

- it is more readable at first sight
- it probably is more newbie-friendly than HTML (and we already use it on GitHub)

rowanthorpe commented:
OK, I have to go out for a while, but when I get back I'll convert the jsdoc-formatting commit to markdown.

rowanthorpe commented:
I rewrote the jsdoc comments in markdown, and added a jsdoc conf-file (needed to activate markdown & to recognize `.gs` extension anyway). I also added two fixup commits at the end.

baatochan commented on code.gs:
```diff
@@ -1588,6 +1587,7 @@ function getEventsOnDate (eventDate, calendarId) {
 }
 
 /**
+ * <div style="clear:both"></div>
```
Makes me wonder why markdown formatting uses html tag...

rowanthorpe replied:
Markdown is just a transpiler to html (and allows inline html) anyway, and from what I could see the standard jsdoc style (for some crazy reason) has the descriptions "clear: none" after the function-name headings, yet a huge right-side padding for the heading. In most cases that is not visible because the function-name/retvals are wide enough to force a linebreak in the formatting anyway. It just happens that the `main()` function has a short name and no retval, so the description ends up being some squished semi-lines far over on the right of it. This ugly clear-both div bugs me, but was the only way I found to fix the formatting without digging into the jsdoc-stylesheet itself (which I really am not in the mood for). If anyone finds a less ugly workaround (or wants to try overriding the stylesheet - perhaps in the conf-file) please add it.

baatochan replied:
Oh, I see, thanks for explaining it for me.

GioBonvi commented:
I have pushed some changes to giobonvi-development which had to be applied on some old commits: this caused conflicts with this PR as I force-pushed.

You might want to rebase on the new version of the branch: moreover given the changes I have made:

- 3d84541 can be removed
- ea1df7, e31c3a8 and e31c3a8 and 7bc0346can be squashed in a secondary-JSDoc-fixes commit
- 2cbcaad might need to be extended as some new classes and functions have been introduced

Let me know what you prefer to do.

rowanthorpe commented:
@giobonvi I did the changes you asked in your last comment, and while I was waiting/rebasing on your changing branch I got a bit lost down the rabbit-hole and ended up adding various commits for cleanups/tweaks/and-lots-of-things, particularly documentation. Now your branch seems to have settled I have rebased-pushed this one. Let me know if anything is unclear.

rowanthorpe commented:
...and I did all the semistandard/CommonMark/JSDoc syntax testing and Javascript/Markdown/JSDoc semantic testing, as now outlined in the CONTRIBUTING file updated by these commits. The syntax tests passed (except for the dangling commas issue which we can't eslint-exempt due to the Google bug), and the semantic tests all looked fine too.

baatochan commented:
Speaking of the dangling commas issue we could check if after the latest Google changes to permission handling it haven't been solved.

rowanthorpe commented:
I very much doubt that would have suddenly changed, as the two still-open issues for it that I mentioned [here](https://github.com/GioBonvi/GoogleContactsEventsNotifier/issues/68#issuecomment-331479039) are from mid-2014 and early-2016 respectively. It's obviously way down their priority lists.

GioBonvi commented on .gitignore:
```diff
@@ -1,2 +1,3 @@
 *~
 /code-customized.gs
+/out/
```
Wouldn't using a more specific name for the JSDoc output folder (like `jsdoc-out`) be more future-proof? I imagine "out" could be the default output folder for many different tools and this would cause conflicts.

rowanthorpe replied:
Yes, easy to change. Was just initially going with the default output-name of the jsdoc tool for the sake of not procrastinating over the name myself. Will change it in a minute.

rowanthorpe replied:
...just to clarify my wording - `...for the sake of not procrastinating over the name myself, if it is better that you decide it anyway...` (wasn't implying that you deciding such things during review is procrastination, rather that I was in danger of getting lost down the rabbit-hole if I focused too deeply on such things at 4 in the morning).

GioBonvi commented on .github/CONTRIBUTING.md:
```diff
@@ -136,6 +136,11 @@ This is the optimal workflow that should be followed when managing a new PR:
+    and it will be a noticeable length of time before that branch is merged to
+    `master`, then the `solved` tag should be added to the PR to make it clear at
+    a glance that nothing else needs doing for that PR in order for it to
+    auto-close.
```
Didn't we verify in our email exchange that PRs get automatically closed when the code is merged in their target branch and not when merged in the default/master branch?

In fact when I added the "solved" label I thought of using it together with the "bug" label marking:

- bug-tracking issues once the bug has been solved (in other words "solved" is the opposite of "wont-fix")
- PRs solving a bug

rowanthorpe replied:
Oops, that was just a wording error due to late-night lack of caffeine. I meant to say that point for **issues** (which don't auto-close until stuff hits default branch), not for **PRs** (which do). I was inspired by seeing all the issues presently sitting open until `development` is merged into `master` (you probably noticed I tagged a lot of them `solved`). Because you prefer that `wont-fix` is mutually exclusive with `solved` (as opposed to being an additional "solution-detail") then I will change that wording: e.g.:

> ...then either the `solved` or `wont-fix` label should be added to the issue...

and will untag the present issues with `wont-fix`. Will do that in a minute.

GioBonvi replied:
Regarding the PR/issue mismatch: ok

Regarding the labels I'd use "solved" as I explained, while we could introduce a "waiting-merge" or something similar to do what you describe.

rowanthorpe replied:
...but unless I've misunderstood, there would then never be a need for a separate "waiting-merge" tag, because:
* a PR gets auto-closed on merge, so it will never need to be tagged "solved" because it will already be closed (and semantically, the PR is never "solved", issues are "solved" by commits/PRs)
* an issue will get auto-closed when a PR which "solves" it is merged to master (so it would make sense to label the issue "solved" when the PR is still only merged to target-branch)
* issues will always therefore be either "unsolved", "wontfix", or "solved" - making "waiting-merge" redundant

rowanthorpe replied:
...of course "unsolved" is not a tag, but is the lack of a "solved" or "wontfix" tag.

GioBonvi replied:
Yes, you're right.  
So we have these "label names": meaning associations:

- "bug": this issue/PR regards a bug in the code
- "solved": the bug this issue/PR is about has been solved (if the issue itself is closed this means the code is now merged into master, otherwise the code is waiting to be merged into master)
- "wontfix": the bug in this issue/PR is about cannot/will not be resolved
- neither "solved" nor "wontfix": the bug this issue/PR is about is still being worked on

rowanthorpe commented:
Those two changes ended up affecting various commits (documentation, functionality, etc). I pushed all changes for them now.

GioBonvi commented on .jsdoc-conf.json:
```diff
@@ -0,0 +1,17 @@
+        "cleverLinks": false,
+        "monospaceLinks": false
+    }
+}
```
Given that we set a default output folder name in the `.gitignore` we could add it here as well:
```
    "opts": {
        "destination": "./jsdoc-out/"
    },
```

I don't know if there are any other useful default options we could set here.

rowanthorpe replied:
Nice :-) I hadn't read the docs for the config-file. `jsdoc-out` was already in the `excludePattern` to avoid any unexpected side-effects. Will add it in a minute (and will update the docs as the optflag now won't be needed by default).

rowanthorpe replied:
...I'm still deciding whether to also add `pedantic`, `verbose`, and `recurse` to the config-file (and remove them as manual flags in the docs), because it could be argued either way whether they are sane **default** behaviour. What do you think? If you can't reply at the moment I will leave it as-is, and that can always be done in a later commit/PR.

GioBonvi commented on .github/CONTRIBUTING.md:
````diff
@@ -77,24 +77,106 @@ Pull Requests complies with these rules:
+    commandline tool using (for example, on a Unix-like system):
+    ```sh
+    npm install -g "markdownlint-cli" # if not yet installed
+    find . -name "*.md" -exec markdownlint {} \;
````
I'd add `-c ".markdownlint.json"` here.

rowanthorpe replied:
Oops, forgot the flag. Will add, thanks.

rowanthorpe commented:
Latest requested changes done (I didn't add extra opts - mentioned in my last comment - to `jsdoc-conf.json`, @GioBonvi if you care about that feel free to do a commit, or let me know if you want me to do a PR/commit).

GioBonvi commented:
I think it's all good the way it is now: I am going to do a last cursory check of the code and then I'll merge.

rowanthorpe commented:
I just discovered you can add `no:assignee` when searching open issues, so I will update the commit with the "find open issues" link in CONTRIBUTING, and force-push it. In case you are already final-reviewing the PR, I promise that is all I will change here.

GioBonvi commented:
@baatochan  
> Speaking of the dangling commas issue we could check if after the latest Google changes to permission handling it haven't been solved.

I forgot to mention this: I checked and the bug is still there.
